### PR TITLE
Signup and Login api routes

### PR DESCRIPTION
**Pull request by gary-bains**
Hi @megangilpin I have moved the DB connection logic to it's own file.

### CODE REVIEW

**Comment by megangilpin** [comment]
I am totally fine if the DB connection is in a different file.  For my future reference is there a reason you moved it? ie: speed, readability or something else.

**Comment by gary-bains** [comment]
> I am totally fine if the DB connection is in a different file. For my future reference is there a reason you moved it? ie: speed, readability or something else.

Well, conventionally the only thing that gets exported out of the `app.js` file is the instance of app itself and nothing else. But we do require connection object in various places, like controllers, so we have to export it from whichever file it is in. To me creating a separate file for this seemed the easiest solution.

File: server/controllers/usersController.js [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
 
+const SALT_ROUNDS = Number(process.env.SALT_ROUNDS);
+
+const create = async ({ email, password }) => {
+    const hashedPassword = await bcrypt.hash(password, SALT_ROUNDS);
+    const user = await User.create({ email, password: hashedPassword });
+    return user;
 };
+
+const findOneWithemail = async (email) => await User.findOne({ email });
```
Comment (ThinkTink) [inline server/controllers/usersController.js]: consistent camelCase would be findOneWithEmail

File: server/controllers/usersController.js [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
-// Defining methods for the usersController
-module.exports = {
-  // define methods here
+const User = connection.model("User", userSchema);
 
+const SALT_ROUNDS = Number(process.env.SALT_ROUNDS);
+
+const create = async ({ email, password }) => {
+    const hashedPassword = await bcrypt.hash(password, SALT_ROUNDS);
+    const user = await User.create({ email, password: hashedPassword });
```
Comment (ThinkTink) [inline server/controllers/usersController.js:12]: We'll need to be using all of fields in Megan's user schema, not just email and password

File: server/middleware.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+const jwt = require("jsonwebtoken");
+
+const userController = require("./controllers/usersController");
```
Comment (ThinkTink) [inline server/middleware.js]: Is this needed here?

File: server/routes/auth.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+});
+
+function validationMiddleware(req, res, next) {
+    const { email, password } = req.body;
+    // input validation
+    const errors = [];
+    if (!email) errors.push("Missing email");
+    if (!password) errors.push("Missing password");
+    if (email && typeof email !== "string") errors.push("Invalid type");
+    if (password && typeof password !== "string") errors.push("Invalid type");
```
Comment (ThinkTink) [inline server/routes/auth.js]: Should we specify in the error message which parameter has the invalid type?

File: server/routes/auth.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+const router = express.Router();
+
+router.post("/login", validationMiddleware, async function (req, res, next) {
+    try {
+        const { email, password } = req.body;
+        // verify that there is user for email
+        const user = await userController.findOneWithemail(email);
+        if (!user) {
+            res.status(400).json({ errors: ["Email or password invalid"] });
```
Comment (ThinkTink) [inline server/routes/auth.js:14]: Let's make the error message as specific as possible - 'No user exists for that email' or something

File: server/controllers/usersController.js [hunk 6]
```diff
@@ -1,9 +1,21 @@
...
-const db = require("../models/user");
+const connection = require("../dbConnection");
+const userSchema = require("../models/user");
 
-// Defining methods for the usersController
-module.exports = {
-  // define methods here
+const User = connection.model("User", userSchema);
 
+const SALT_ROUNDS = Number(process.env.SALT_ROUNDS);
```
Comment (ThinkTink) [inline server/controllers/usersController.js:8]: We want to make sure everyone is using consistent values for their env, so we'll need to sync about this and all the other env values in this PR at some point

File: server/routes/auth.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            return;
+        }
+        // create and return jwt with user obj
+        const responseObj = await createResponseObj(user._doc);
+        res.cookie("token", responseObj.token, { httpOnly: true });
+        res.json(responseObj);
+    } catch (error) {
+        console.error(error);
+        next(error);
```
Comment (ThinkTink) [inline server/routes/auth.js]: We should probably just return a 400 here, right?

File: server/routes/auth.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    }
+});
+
+function validationMiddleware(req, res, next) {
+    const { email, password } = req.body;
+    // input validation
+    const errors = [];
+    if (!email) errors.push("Missing email");
+    if (!password) errors.push("Missing password");
+    if (email && typeof email !== "string") errors.push("Invalid type");
```
Comment (ThinkTink) [inline server/routes/auth.js]: We want to verify that it's a valid email address format at some point as well

**Review by ThinkTink — COMMENTED** [review]
Good work here! Left mostly cleanup comments.
The big thing left is getting the full user schema from Megan's branch and expanding the create function to account for all of the different fields

File: server/controllers/usersController.js (hunk 2 above)
Comment (megangilpin) [reply server/controllers/usersController.js:12]: I think to make the sign up smoother I should either remove the required on the address or separate the profile info from the email and password and put it into its own profile collection.

File: server/controllers/usersController.js (hunk 1 above)
Comment (gary-bains) [reply server/controllers/usersController.js]: Good catch, I just fixed it.

File: server/routes/auth.js (hunk 4 above)
Comment (gary-bains) [reply server/routes/auth.js]: Agree, Fixed it !

File: server/controllers/usersController.js (hunk 6 above)
Comment (gary-bains) [reply server/controllers/usersController.js:8]: Yes, we discussed the same thing and came agreement to use slack to sync our `.env` files

File: server/routes/auth.js (hunk 5 above)
Comment (gary-bains) [reply server/routes/auth.js:14]: We could, but wouldn't want to give out the hackers easy way to find out which emails/usernames exists in out system. 

Thoughts? @sagar-13 @megangilpin @sina-jamshidi

File: server/routes/auth.js (hunk 8 above)
Comment (gary-bains) [reply server/routes/auth.js]: I added a simple regex for the email validation, we can further improve it in future.

File: server/routes/auth.js (hunk 7 above)
Comment (gary-bains) [reply server/routes/auth.js]: As we aren't sure what went wrong for the code to reach this catch, so I think we should return 500 as it is likely something that went wrong on the server side.

File: server/test/ping.spec.js [hunk 9]
```diff
@@ -6,17 +6,16 @@ chai.should();
...
-      .send({ teamName: "Shums" })
-      .end((err, res) => {
-        res.should.have.status(400);
-        res.body.should.have
-          .property("response")
-          .eql("Shums is not part of the team. Modify your .env");
-        done();
-      });
-  });
+    it("it should return 400", (done) => {
```
Comment (sina-jamshidi) [inline server/test/ping.spec.js]: we can get rid of this entire file

**Review by sina-jamshidi — APPROVED** [review]
Looks good to merge!

File: server/app.js [hunk 10]
```diff
@@ -1,51 +1,44 @@
...
+const mongoose = require("mongoose");
 const createError = require("http-errors");
 const express = require("express");
 const { join } = require("path");
 const cookieParser = require("cookie-parser");
 const logger = require("morgan");
 
+const { loginRequired } = require("./middleware");
+const authRouter = require("./routes/auth");
+const usersRouter = require("./routes/users");
```
Comment (garys-io) [inline server/app.js:11]: cas

**Pull request merged**